src/conditional.ts

```diff
@@ -0,0 +1,198 @@
+ * @signature
+ *   R.conditional(data, ...cases);
+ * @example
+ *   const nameOrId = 3 as string | number;
+ *   R.conditional(
+ *     nameOrId,
+ *     [R.isString, (name) => `Hello ${name}`],
+ *     [R.isNumber, (id) => `Hello ID: ${id}`],
+ *     R.conditional.defaultCase(
+ *       (something) => `Hello something (${JSON.stringify(something)})`,
+ *     ),
+ *   ); //=> 'Hello ID: 3'
+ * @dataFirst
+ * @category Function
+ */
+export function conditional<
+  T,
+  Return0,
+  Return1 = never,
+  Return2 = never,
+  Return3 = never,
+  Return4 = never,
+  Return5 = never,
+  Return6 = never,
+  Return7 = never,
+  Return8 = never,
+  Return9 = never,
+  Thru0 extends T = T,
+  Thru1 extends T = T,
+  Thru2 extends T = T,
+  Thru3 extends T = T,
+  Thru4 extends T = T,
+  Thru5 extends T = T,
+  Thru6 extends T = T,
+  Thru7 extends T = T,
+  Thru8 extends T = T,
+  Thru9 extends T = T,
+>(
+  data: T,
+  case0: Case<T, Return0, Thru0>,
+  case1?: Case<T, Return1, Thru1>,
+  case2?: Case<T, Return2, Thru2>,
+  case3?: Case<T, Return3, Thru3>,
+  case4?: Case<T, Return4, Thru4>,
+  case5?: Case<T, Return5, Thru5>,
+  case6?: Case<T, Return6, Thru6>,
+  case7?: Case<T, Return7, Thru7>,
+  case8?: Case<T, Return8, Thru8>,
+  case9?: Case<T, Return9, Thru9>
+):
+  | Return0
+  | Return1
+  | Return2
+  | Return3
+  | Return4
+  | Return5
+  | Return6
+  | Return7
+  | Return8
+  | Return9;
+
+export function conditional(): unknown {
+  return purryOn(isCase, conditionalImplementation, arguments);
+}
+
+function conditionalImplementation<In, Out>(
+  data: In,
+  ...cases: ReadonlyArray<Case<In, Out>>
+): Out {
+  for (const [when, then] of cases) {
+    if (when(data)) {
+      return then(data);
+    }
+  }
+
+  throw new Error('conditional: data failed for all cases');
+}
+
+function isCase(maybeCase: unknown): maybeCase is Case<unknown, unknown> {
+  if (!Array.isArray(maybeCase)) {
+    return false;
+  }
+
+  const [when, then, ...rest] = maybeCase;
+  return (
+    typeof when === 'function' &&
+    when.length <= 1 &&
+    typeof then === 'function' &&
+    then.length <= 1 &&
+    rest.length === 0
+  );
+}
+
+export namespace conditional {
+  /**
+   * A simplified case that accepts all data. Put this as the last case to prevent an exception from being thrown when none of the previous cases matched. If this is not the last case it will short-circuit anything after it.
+   * @param then - You only need to provide the transformer, the predicate is implicit.
+   */
+  export const defaultCase = <In>(then: (data: In) => unknown) =>
+    [() => true, then] as const;
+
+  /**
+   * An even simpler catch-all case that simply returns undefined, to match the behavior of other frameworks.
+   * @see defaultCase.
+   */
+  export const LEGACY_DEFAULT_CASE = defaultCase(() => undefined);
```

Comment:
just a minor nit: why is this named `LEGACY`? does that mean it shouldn't be used? If so, why are we adding it ?

Reply:
Lodash and Ramda return `undefined` unconditionally because they have to take the final "else" case into account in the typing. I didn't want the return type to include `undefined` if it could be avoided as it could be messy for the rest of the pipe to handle it. I added the "defaultCase" function to make it more ergonomic, but then I wanted to make migration from other frameworks even easier, so I added the default behavior as an even easier drop-in const.

I called it "legacy" because I assumed that people coming from another framework would first use this to transition smoothly without changing behavior, and then they could tackle these cases in a later refactor. For those people, the existing behavior is the legacy/old/previous way things were done; similar to a "_FIX_ME" added after a deprecation or refactor.

I added a note about it in the function documentation:
> Unlike similar implementations in other frameworks, this does not return a default value when no cases match (like `undefined`), throwing an exception instead. To implement traditional default behavior, use `conditional.LEGACY_DEFAULT_CASE` as the final case or `conditional.defaultCast(then)` which allows for customized default handling.

does that make sense? if no, can you suggest an alternative?